functorch/_src/vmap.py

```diff
@@ -366,4 +366,119 @@ def wrapped(*args, **kwargs):
+
+    Args:
+        func (function): A Python function that takes one or more arguments.
+            Must return one or more Tensors.
+        in_dims (int or nested structure): Specifies which dimension of the
+            inputs should be mapped over. :attr:`in_dims` should have a
+            structure like the inputs. If the :attr:`in_dim` for a particular
+            input is None, then that indicates there is no map dimension.
+            Default: 0.
+        out_dims (int or Tuple[int]): Specifies where the mapped dimension
+            should appear in the outputs. If :attr:`out_dims` is a Tuple, then
+            it should have one element per output. Default: 0.
+        randomness (str): Specifies whether the randomness in this
+            vmap should be the same or different across batches. If 'different',
+            the randomness for each batch will be different. If 'same', the
+            randomness will be the same across batches. If 'error', any calls to
+            random functions will error. Default: 'error'. WARNING: this flag
+            only applies to random PyTorch operations and does not apply to
+            Python's random module or numpy randomness.
+        chunks (int): Number of chunks to use to split the input data. Default is 2.
+            If equals to 1 then :func:`vjp` is called.
+
+    Returns:
+        Returns a new "batched" function. It takes the same inputs as
+        :attr:`func`, except each input has an extra dimension at the index
+        specified by :attr:`in_dims`. It takes returns the same outputs as
+        :attr:`func`, except each output has an extra dimension at the index
+        specified by :attr:`out_dims`.
+    """
+    if randomness not in ['error', 'different', 'same']:
+        raise RuntimeError(f"Only allowed values for randomness are 'error', 'different', or 'same'. Got {randomness}")
+
+    if chunks == 1:
+        return vmap(func, in_dims=in_dims, out_dims=out_dims, randomness=randomness)
+
+    def _flat_vmap(batch_size, flat_in_dims, flat_args, args_spec, **kwargs):
+        vmap_level = _vmap_increment_nesting(batch_size, randomness)
+        try:
+            batched_inputs = _create_batched_inputs(flat_in_dims, flat_args, vmap_level, args_spec)
+            batched_outputs = func(*batched_inputs, **kwargs)
+            return _unwrap_batched(batched_outputs, out_dims, vmap_level, batch_size, func)
+        finally:
+            _vmap_decrement_nesting()
+
+    def _get_chunk_flat_args(flat_args_, flat_in_dims_, chunks_):
+        flat_args_chunks = tuple(
+            t.chunk(chunks_, dim=in_dim) if in_dim is not None else [t, ] * chunks_
+            for t, in_dim in zip(flat_args_, flat_in_dims_)
+        )
+        # transpose chunk dim and flatten structure
+        # chunks_flat_args is a list of flatten args
+        chunks_flat_args = zip(*flat_args_chunks)
+        return chunks_flat_args
```

Comment:
I feel like we don't actually need to peek into vmap internals to do this, but I also have not thought about this problem as much as you have. 

Could we implement chunk_vmap like the following?
```py
def chunk_vmap(func, in_dims, out_dims, randomness, chunks):
  if chunks == 1:
    return vmap(func, in_dims, out_dims, randomness)
  def inner(*args):
    chunked_args = chunk_args(args)
    chunked_outputs = [vmap(func, in_dims, out_dims, randomness)(args) for args in chunked_args]
    outputs = concat_chunks(chunked_outputs)
    return outputs
  return inner
```
This way if someone changes how vmap works (e.g. adds a new randomness option), we do not have to change it inside vmap and here

Reply:
Right now the implementation is almost like you described but uses flat_vmap instead of vmap as it is more simple to chunk flat args.
If doing as you suggest:
```python
    chunked_args = chunk_args(args)
    chunked_outputs = [vmap(func, in_dims, out_dims, randomness)(args) for args in chunked_args]
```
Behind the scene, we may probably flatten and unflatten 2 times more the args as `chunk_args` will flatten, chunk, unflatten and vmap will also flatten, flat_vmap, unflatten.

If flatten/unflatten is not a problem we can try to code it according to you suggestion.

Reply:
> If flatten/unflatten is not a problem

I see, thanks for the clarification. I agree, it would be nice to avoid too many flattens and unflattens. Let's stick with your approach; I will add some suggestions so that we repeat ourselves less.